synapse/lib/snap.py

```diff
@@ -1388,6 +1392,132 @@ async def addNode(self, name, valu, props=None, norminfo=None):
+                norm, norminfo = prop.type.norm(valu)
+
+                norms[name] = (prop, norm, norminfo)
+                proplist.append((name, norm))
+        except s_exc.BadTypeValu as e:
+            if not trycast: raise
+            await self.warn(f'Bad type value: {e}')
+            return
+
+        proplist.sort()
+
+        # check first for an exact match via our same deconf strategy
+        iden = s_common.guid(proplist)
+
+        node = await self.getNodeByNdef((form.full, iden))
+        if node is not None:
+
+            # ensure we still match the property deconf criteria
+            match = True
+            for name, (prop, norm, info) in norms.items():
+                if not self._filtByPropAlts(node, prop, norm):
+                    match = False
+                    break
+
+            if match:
+                # ensure the non-deconf props are set
+                async with self.getEditor() as editor:
+                    proto = editor.loadNode(node)
+                    for name, (valu, info) in props.items():
+                        await proto.set(name, valu, norminfo=info)
+                return node
+
+        # TODO there is an opportunity here to populate
+        # a look-aside for the alternative iden to speed
+        # up future deconfliction and potentially pop them
+        # if we lookup a node and it no longer passes the
+        # filter...
+
+        # no exact match. lets do some counting.
+        for name, (prop, norm, info) in norms.items():
+            count = await self._getPropAltCount(prop, norm)
+            counts.append((count, prop, norm))
+
+        counts.sort()
+
+        # lift starting with the lowest count
+        count, prop, norm = counts[0]
+        async for node in self._nodesByPropAlts(prop, norm):
+            await asyncio.sleep(0)
+
+            # filter on the remaining props/alts
+            match = True
+            for count, prop, norm in counts[1:]:
+                if not self._filtByPropAlts(node, prop, norm):
+                    match = False
+                    break
+
+            if match:
+                # ensure the non-deconf props are set
+                async with self.getEditor() as editor:
+                    proto = editor.loadNode(node)
+                    for name, (valu, info) in props.items():
+                        await proto.set(name, valu, norminfo=info)
+                return node
+
+        async with self.getEditor() as editor:
+            proto = await editor.addNode(form.name, iden)
+            for name, (prop, valu, info) in norms.items():
+                await proto.set(name, valu, norminfo=info)
+            for name, (valu, info) in props.items():
+                await proto.set(name, valu, norminfo=info)
+
+        return await self.getNodeByBuid(proto.buid)
+
+    async def _getPropAltCount(self, prop, valu):
+        count = 0
+        for prop in prop.getAlts():
+            if prop.type.isarray:
```

Comment:
I think these isarray checks also need to verify that `valu` is not an array, otherwise we still want to use `getPropCount` to count full array matches

Reply:
Would apply to `_filtByPropAlts` and `_nodesByPropAlts` as well

Reply:
Doesn't necessarily have to be an alts, this would apply to any array property you wanted to specify all the values for, like:
```
[ transport:air:flight=({"departed": "2020", "stops": ["foo", "bar"]}) ]
```